Решение 7 спринта

Mexaelb opened the pull request:
итоговая работа по 7 спринту

Commit:
тест ручки api/v1/courier

добавил readme
добавил генератор пользователей
тестовые данные
тесты для ручки api/v1/courier

Commit:
добавил тесты на логин пользователя

Commit:
доработки

добавил директорию с отчетами тестов allure
добавил фикстуры
добавил тест апи логина курьера
добавил параметризацию

Commit:
директория с отчетами по тестам

Commit:
правки

удалил лишний файл
добавились отчеты по тестам
добавил урлы в переменные
обновил тесты

Commit:
доработки

тест на создание заказа

Commit:
исправил тест

исправил тест , сделал параметризацию

Commit:
исправил тест заказа добавил фикстуру

Commit:
добавил тест на получение списка заказов

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,40 @@
+    return payload
+
+
+@pytest.fixture
```
Нужно исправить: фикстуры не занимаются прокидыванием данных в тест и выполнением примитивной логики, они выполняют сложную логику предусловий\постусловий и вычислений. 
Статичные данные хранятся в data и просто импортируются куда необходимо

AlexKitIt commented on test/test_create_courier.py:
```diff
@@ -0,0 +1,84 @@
+        response2 = requests.post(f"{Urls.base_url}{Urls.api_create_courier}", data=create_courier_dto)
+
+        assert response1.status_code == 201
+        assert response2.status_code == 409
```
Нужно исправить: здесь и далее: одной проверки на статус код недостаточно, необходимо парсить ответ

Mexaelb replied:
исправлено

AlexKitIt commented on test/test_create_courier.py:
```diff
@@ -0,0 +1,84 @@
+    @allure.title('Проверка создание курьера')
+    def test_create_courier(self,create_courier_dto):
+
+        response = requests.post(f"{Urls.base_url}{Urls.api_create_courier}", data=create_courier_dto)
```
Нужно исправить: здесь и далее: по условию задания все созданные сущности должны быть удалены после теста. Это стоит реализовывать в фикстурах

AlexKitIt commented on user_generate.py:
```diff
@@ -0,0 +1,40 @@
+# если регистрация не удалась, возвращает пустой список
+def register_new_courier_and_return_login_password():
+    # метод генерирует строку, состоящую только из букв нижнего регистра, в качестве параметра передаём длину строки
+    def generate_random_string(length):
+        letters = string.ascii_lowercase
+        random_string = ''.join(random.choice(letters) for i in range(length))
+        return random_string
+
+    # создаём список, чтобы метод мог его вернуть
+    login_pass = []
+
+    # генерируем логин, пароль и имя курьера
+    login = generate_random_string(10)
+    password = generate_random_string(10)
+    first_name = generate_random_string(10)
+
+
+    # собираем тело запроса
+    payload = {
+        "login": login,
+        "password": password,
+        "firstName": first_name
+    }
```
Нужно исправить: копипаста из conftest. Реализация должна быть одна. Если это нужно отдельно - выносим в отдельный метод

AlexKitIt commented on user_generate.py:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    # отправляем запрос на регистрацию курьера и сохраняем ответ в переменную response
+    response = requests.post('https://qa-scooter.praktikum-services.ru/api/v1/courier', data=payload)
```
Нужно исправить: урл вынесен в test_data

Mexaelb replied:
исправлено

AlexKitIt commented on test/test_login_courier.py:
```diff
@@ -0,0 +1,83 @@
+    @allure.title('курьер может авторизоваться')
+    def test_login_courier(self):
+
+        login_pass = register_new_courier_and_return_login_password()
```
Нужно исправить: здесь и далее: в тестах логина это предусловие, оно должно быть описано в фикстуре

Commit:
вынес урл в test_data

Commit:
в каждый тест добавил проверку ответа

Commit:
поправил модуль с тестовыми данными

вынес из conftest в testdata подготовку данных для заказа
исправил тест проверки создания заказа

Commit:
исправил тесты

в test_data перенес подготовку данных для создания курьера
исправил тесты

Commit:
правки

Commit:
правки

Commit:
правки

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,10 @@
+    courier = CourierClass()
+    courier.register_new_courier_and_return_login_password()
+    yield courier
+    courier.delete_courier(id)
```
Нужно исправить: какой id тут передается? Где он берется?

Mexaelb replied:
исправлено

AlexKitIt commented on src/courier.py:
```diff
@@ -0,0 +1,74 @@
+    def login_courier(self):
+
+        login_pass = CourierClass()
+        f = login_pass.register_new_courier_and_return_login_password()
```
Можно лучше: нейминги необходимо выбирать осознанные

AlexKitIt commented on src/courier.py:
```diff
@@ -0,0 +1,74 @@
+    @allure.step('авторизация курьера')
+    def login_courier(self):
+
+        login_pass = CourierClass()
```
Нужно исправить: здесь и далее: создание объектов класса внутри этого же класса не допускается. Обращение к методам класса из этого класса ведется через self

Mexaelb replied:
исправлено

AlexKitIt commented on test/test_login_courier.py:
```diff
@@ -0,0 +1,62 @@
+
+class TestLoginCourier:
+
+    @allure.title('курьер может авторизоваться')
+    def test_login_courier(self):
+
+        login_pass = CourierClass()
+        status_code, response = login_pass.login_courier()
+        assert status_code == 200 and 'id' in response
+
+        courier_id = response['id']
+        login_pass.delete_courier(courier_id)
+
+
+    @allure.title('для авторизации нужно передать все обязательные поля')
+    def test_login_courier_all_fields(self):
+
+        login_pass = CourierClass()
+        status_code, response = login_pass.login_courier()
+        assert status_code == 200 and 'id' in response
+
+        courier_id = response['id']
+        login_pass.delete_courier(courier_id)
```
Нужно исправить: эти два и test_login_courier_ok_id копипаста. Это один тест, почему тут их стало три?

Mexaelb replied:
исправил

AlexKitIt commented on test_data.py:
```diff
@@ -0,0 +1,55 @@
+
+        return payload
+
+    @staticmethod
+    @allure.step('создание логин/пароль/имя курьера')
+    def login_pass_name_courier_dto():
+        def generate_random_string(length):
+            letters = string.ascii_lowercase
+            random_string = ''.join(random.choice(letters) for i in range(length))
+            return random_string
+
+        login = generate_random_string(10)
+        password = generate_random_string(10)
+        first_name = generate_random_string(10)
+
+        payload = {
+            "login": login,
+            "password": password,
+            "firstName": first_name
+        }
+
+        return payload
```
Нужно исправить: методы с логикой живут не в data а в helpers

Mexaelb replied:
исправлено

Commit:
правки

Commit:
правки

Commit:
правки

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,10 @@
+from src.courier import CourierClass
+
+@pytest.fixture()
+def courier_t():
```
Нужно исправить: фикстура написана но нигде не используется

Mexaelb replied:
удалил

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,10 @@
+    courier = CourierClass()
+    courier.register_new_courier_and_return_login_password()
+    yield courier
+    courier.delete_courier()
```
Нужно исправить: метод этого класса принимает courier_id, тут он не передается

Mexaelb replied:
conftest.py удалил

AlexKitIt commented on src/courier.py:
```diff
@@ -0,0 +1,72 @@
+    @allure.step('авторизация курьера')
+    def login_courier(self):
+
+        f = self.register_new_courier_and_return_login_password()
```
Нужно исправить: здесь и далее: после всех этих созданий курьеров ни один не удаляется, метод логина должен отвечать за логин а не за регистрацию и логи

Mexaelb replied:
исправлено

AlexKitIt commented on test/test_create_courier.py:
```diff
@@ -0,0 +1,92 @@
+    @allure.title('Проверка создание курьера')
+    def test_create_courier(self):
+
+        login_pass = CourierClass()
+        status_code, response = login_pass.register_new_courier_and_return_lp()
+
+        assert response == {'ok': True} and status_code == 201
```
Нужно исправить: здесь и далее: во всех этих тестах курьер не удаляется после теста

Mexaelb replied:
исправлено

Commit:
удалил conftest.py

Commit:
Revert "удалил conftest.py"

This reverts commit e4542674a13642fafc4d3a1698c9e5d58f7f48be.

Commit:
правки

переписана фикстура - теперь используется
переписаны тесты авторизации
переписаны тесты создания курьера
вынесены в фикстуру создание и удаление тестовых данных для курьера
описаны helpers
